crates/acvm/src/pwg/binary.rs

```diff
@@ -0,0 +1,236 @@
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+use acir::{
+    circuit::{gate::Directive, Gate},
+    native_types::{Expression, Witness},
+    FieldElement,
+};
+use num_bigint::BigUint;
+use num_traits::{One, Zero};
+
+use crate::GateResolution;
+
+// This structure tracks un-resolved witnesses which are constrained to be booleans or sum of booleans, so that gates which uses only those can be solved.
+// This is the case for array equality constraints which are unsolved when the value of arrays returned from main are not supplied in the .toml
+pub struct BinarySolver {
+    binary_witness: HashSet<Witness>,
+    invert_witness: HashMap<Witness, Witness>,
+    positive_witness: HashMap<Witness, BigUint>,
+}
+
+impl Default for BinarySolver {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BinarySolver {
+    pub fn new() -> BinarySolver {
+        BinarySolver {
+            binary_witness: HashSet::new(),
+            invert_witness: HashMap::new(),
+            positive_witness: HashMap::new(),
+        }
+    }
+
+    pub fn is_boolean(&self, w: &Witness) -> bool {
+        self.binary_witness.contains(w)
+    }
+
+    pub fn are_inverse(&self, w1: &Witness, w2: &Witness) -> bool {
+        self.invert_witness.get(w1) == Some(w2) || self.invert_witness.get(w2) == Some(w1)
+    }
+
+    pub fn are_boolean(&self, w1: &Witness, w2: &Witness) -> bool {
+        self.are_inverse(w1, w2) || (self.is_boolean(w1) && self.is_boolean(w2))
+    }
+
+    pub fn get_max_value(&self, w: &Witness) -> Option<BigUint> {
+        if self.is_boolean(w) {
+            Some(BigUint::one())
+        } else {
+            self.positive_witness.get(w).cloned()
+        }
+    }
+
+    pub fn solve(
+        &mut self,
+        gate: &Gate,
+        initial_witness: &mut BTreeMap<Witness, FieldElement>,
+    ) -> GateResolution {
+        let mut result = GateResolution::Skip;
+        if let Gate::Arithmetic(arith) = gate {
+            let partial_gate =
+                super::arithmetic::ArithmeticSolver::evaluate(arith, initial_witness);
+            result = self.solve_booleans(initial_witness, &partial_gate);
+            self.identify_booleans(&partial_gate);
+        } else {
+            self.identify_binaries(gate);
+        }
+        result
```

Comment:
```suggestion
        if let Gate::Arithmetic(arith) = gate {
            let partial_gate =
                super::arithmetic::ArithmeticSolver::evaluate(arith, initial_witness);
            let result = self.solve_booleans(initial_witness, &partial_gate);
            self.identify_booleans(&partial_gate);
            result
        } else {
            self.identify_binaries(gate);
            GateResolution::Skip
        }
```
We can remove some indirection with the mutable result here.